### PR TITLE
fix: Omnichannel Tags available to be used in the wrong department

### PR DESCRIPTION
**Pull request by murtaza98**
## Issue(s)

https://rocketchat.atlassian.net/browse/OC-352
## Further comments

- [x] Add API tests for new departments filter

### CODE REVIEW

**Review by ggazzo — APPROVED** [review]

**Review by MartinSchoeler — APPROVED** [review]

**Comment by murtaza98** [comment]
Hi @MartinSchoeler @ggazzo I think we'll need your approval here again! Lost your previous approval with recent changeset commit. Thanks!

**Review by MartinSchoeler — APPROVED** [review]

File: apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts:42]: This still allows users to see tags from other departments just by knowing the id 😬 we should add a check of user's departments and then return the tag list of _only_ the departments user can see

Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts:42]: Ouch! I missed that! We'd need to address this before merging this PR.

Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts]: image

It should be fixed now

File: apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts:71]: hmmm, but `$and` is not optional if you're not typing it as optional. Maybe if you change the type of `query` to be `Filter<x>` would it be optional, then you can use `DeepWritable` to remove the readonly status of each type

File: apps/meteor/ee/app/livechat-enterprise/server/api/lib/departments.ts
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/api/lib/departments.ts:30]: You can create a new query to check the department is on the scope for the monitor instead of fetching all the ones it can view 👀

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts:71]: Yups, I tried that but it was getting messy dealing with those types, so I went with this. But let me tried to give it another try and if you have suggestions please feel free to share :v:

Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts]: found an elegant solution to :partying_face:

File: apps/meteor/ee/app/livechat-enterprise/server/api/lib/departments.ts
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/lib/departments.ts]: Good idea! This should be implemented now. Pls check

**Comment by murtaza98** [comment]
Hello @KevLehman @MartinSchoeler @ggazzo Could you please review this PR once again? Appreciate your help here :pray: Thanks!

**Review by KevLehman — APPROVED** [review]

**Review by MartinSchoeler — APPROVED** [review]

File: apps/meteor/client/components/Omnichannel/Tags.tsx
Comment (ggazzo) [inline apps/meteor/client/components/Omnichannel/Tags.tsx:34]: what about the `department` prop? its being ignored from the key

**Review by ggazzo — DISMISSED** [review]

**Review by KevLehman — APPROVED** [review]

File: apps/meteor/client/components/Omnichannel/Tags.tsx
Comment (murtaza98) [reply apps/meteor/client/components/Omnichannel/Tags.tsx]: Valid point. Let's do it

**Review by KevLehman — APPROVED** [review]

**Review by KevLehman — DISMISSED** [review]

**Review by MartinSchoeler — DISMISSED** [review]

**Review by KevLehman — APPROVED** [review]

**Review by MartinSchoeler — APPROVED** [review]

**Review by ggazzo — APPROVED** [review]

**Pull request merged**
